Add refs for torch.unfold and a decomposition for its backward.

lezcano opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #85629

It's not clear to me what's the difference between `unfold` and `unfold_copy`, as this latter one is codegen'd

I also took this chance to clean the implementation of unfold and its reference

lezcano committed:
Add refs for torch.unfold and a decomposition for its backward.

It's not clear to me what's the difference between `unfold` and `unfold_copy`, as this latter one is codegen'd

I also took this chance to clean the implementation of unfold and its reference

[ghstack-poisoned]

lezcano committed:
Update on "Add refs for torch.unfold and a decomposition for its backward."


It's not clear to me what's the difference between `unfold` and `unfold_copy`, as this latter one is codegen'd

I also took this chance to clean the implementation of unfold and its reference

[ghstack-poisoned]

lezcano committed:
Update on "Add refs for torch.unfold and a decomposition for its backward."


It's not clear to me what's the difference between `unfold` and `unfold_copy`, as this latter one is codegen'd

I also took this chance to clean the implementation of unfold and its reference

[ghstack-poisoned]

mruberry commented on aten/src/ATen/native/TensorShape.cpp:
```diff
@@ -3302,35 +3302,22 @@ Tensor detach(const Tensor& self) {
-
-  new_size[self.dim()] = size;
-  new_stride[self.dim()] = self.dim() == 0 ? 1 : strides[dimension];
-  for(const auto d : c10::irange(self.dim())) {
```
What a crazy for-loop

ngimel commented on torch/_refs/__init__.py:
```diff
@@ -3565,12 +3555,21 @@ def transpose(a: TensorLikeType, dim0: int, dim1: int) -> TensorLikeType:
+@register_decomposition(torch.ops.aten.unfold_copy)
+@out_wrapper()
+def unfold_copy(self: TensorLikeType, dimension: int, size: int, step: int):
+    # nb. It's not clear to me what's the difference between this function and torch.unfold
```
unfold_copy returns a contiguous tensor with the result, whereas `unfold` returns restrided view of the input. It's unfortunate view tests no longer catch those discrepancies.

mruberry replied:
The signatures in native_functions.yaml show this aliasing relationship:

https://github.com/pytorch/pytorch/blob/29c78266c046c7f83e7d84fc764af47e62ae9542/aten/src/ATen/native/native_functions.yaml#L8881

https://github.com/pytorch/pytorch/blob/29c78266c046c7f83e7d84fc764af47e62ae9542/aten/src/ATen/native/native_functions.yaml#L12925

There is a test that unfold returns a view (at least in one case):

https://github.com/pytorch/pytorch/blob/0b93afb112d48bb6d89a1e183a90b403560e84e4/test/test_view_ops.py#L602

```
t = torch.ones(10)
v = t.unfold(0, 3, 2)
v._is_view()
: True

t = torch.ones(10)
c = torch.unfold_copy(t, 0, 3, 2)
c._is_view()
: False
```

ngimel replied:
Right, there are one-off tests here and there, but in general we've given up on testing that refs preserve view relationships of original ops, and refs won't be tested on those one-off tests.

mruberry replied:
Correct, sorry, didn't mean to suggest otherwise

bdhirsh replied:
> Ok, so hear this out. It so happens that unfold_copy does return a view lol

Do you have a repro? The impl for the view copy kernels lives in `build/aten/src/ATen/CompositeViewCopyKernel.cpp` (codegen), and looks like this:
```
at::Tensor unfold_copy(const at::Tensor & self, int64_t dimension, int64_t size, int64_t step) {
  auto output = at::_ops::unfold::call(self, dimension, size, step);
    return output.clone();
}
```

The output isn't a view (since we call `.clone()`). Although the output should be contiguous, as Natalia mentioned above, and that technically isn't always the case. That should be fixed by https://github.com/pytorch/pytorch/pull/85747.

What's the failing out= test? My guess is that the out= kernels aren't resizing properly (they're implemented today by just calling the original `unfold_copy()` op, followed by a `copy_()` to the out= tensor).

ngimel replied:
I don't think unfold_copy produces a view, except in rare circumstances where unfold is a no-op:
```
In [3]: x = torch.arange(1., 8)

In [4]: x.data_ptr()
Out[4]: 93952253159872

In [5]: x.unfold(0, 2, 1).data_ptr()
Out[5]: 93952253159872

In [6]: x.unfold(0, 2, 1).is_contiguous()
Out[6]: False

In [11]: out=torch.unfold_copy(x, 0, 2, 1)

In [12]: out.data_ptr()
Out[12]: 93952337762432

In [13]: out.is_contiguous()
Out[13]: True
```

lezcano replied:
oof. Sorry for the noise. I should have done an unconditional copy, but `contiguous()` sometime returns an alias. That's fixed.

Now, the `out=` errors are real. I have not investigated what's it that's failing exactly.

lezcano committed:
Update on "Add refs for torch.unfold and a decomposition for its backward."


It's not clear to me what's the difference between `unfold` and `unfold_copy`, as this latter one is codegen'd

I also took this chance to clean the implementation of unfold and its reference

[ghstack-poisoned]

lezcano committed:
Update on "Add refs for torch.unfold and a decomposition for its backward."


It's not clear to me what's the difference between `unfold` and `unfold_copy`, as this latter one is codegen'd

I also took this chance to clean the implementation of unfold and its reference

[ghstack-poisoned]

lezcano committed:
Update on "Add refs for torch.unfold and a decomposition for its backward."


It's not clear to me what's the difference between `unfold` and `unfold_copy`, as this latter one is codegen'd

I also took this chance to clean the implementation of unfold and its reference

[ghstack-poisoned]

lezcano committed:
Update on "Add refs for torch.unfold and a decomposition for its backward."


It's not clear to me what's the difference between `unfold` and `unfold_copy`, as this latter one is codegen'd

I also took this chance to clean the implementation of unfold and its reference

[ghstack-poisoned]

lezcano committed:
Update on "Add refs for torch.unfold and a decomposition for its backward."


It's not clear to me what's the difference between `unfold` and `unfold_copy`, as this latter one is codegen'd

I also took this chance to clean the implementation of unfold and its reference

[ghstack-poisoned]

lezcano committed:
Update on "Add refs for torch.unfold and a decomposition for its backward."


It's not clear to me what's the difference between `unfold` and `unfold_copy`, as this latter one is codegen'd

I also took this chance to clean the implementation of unfold and its reference

[ghstack-poisoned]

lezcano committed:
Update on "Add refs for torch.unfold and a decomposition for its backward."


It's not clear to me what's the difference between `unfold` and `unfold_copy`, as this latter one is codegen'd

I also took this chance to clean the implementation of unfold and its reference

[ghstack-poisoned]

lezcano committed:
Update on "Add refs for torch.unfold and a decomposition for its backward."


It's not clear to me what's the difference between `unfold` and `unfold_copy`, as this latter one is codegen'd

I also took this chance to clean the implementation of unfold and its reference

[ghstack-poisoned]

lezcano committed:
Update on "Add refs for torch.unfold and a decomposition for its backward."


It's not clear to me what's the difference between `unfold` and `unfold_copy`, as this latter one is codegen'd

I also took this chance to clean the implementation of unfold and its reference

[ghstack-poisoned]

lezcano commented:
@mruberry @ngimel this one's ready for review.

mruberry commented on aten/src/ATen/native/TensorShape.cpp:
```diff
@@ -3302,35 +3302,25 @@ Tensor detach(const Tensor& self) {
-  TORCH_CHECK(size <= max_size, "maximum size for tensor at dimension ", dimension,
+Tensor unfold(const Tensor& self, int64_t d, int64_t size, int64_t step) {
+  // some special handling to deal with allow d == 0 when self.dim() == 0
+  auto ndim = self.dim();
```
Nit: `const auto`

lezcano replied:
I used to use `const` everywhere, but in the end I got bored. To the eyes of the compiler, `auto` already deduces `const`. Now, I think Rust's way is cooler in this regard: const by default! Rust, what a cool language.

mruberry reviewed (approved):
Cool!

lezcano committed:
Update on "Add refs for torch.unfold and a decomposition for its backward."


It's not clear to me what's the difference between `unfold` and `unfold_copy`, as this latter one is codegen'd

I also took this chance to clean the implementation of unfold and its reference

[ghstack-poisoned]

lezcano commented:
@pytorchbot merge

lezcano commented:
@pytorchbot merge -f "I think the failure is just coming from a flaky test"

eellison commented on test/test_ops.py:
```diff
@@ -1815,6 +1815,8 @@ def test_refs_are_in_decomp_table(self, op):
     "nn.functional.huber_loss",
     "nn.functional.logsigmoid",
     "nn.functional.multilabel_soft_margin_loss",
+    "unfold",  # fails an aliasing assert
```
In the future, could we fix these asserts before landing ? Or tag me on them ? cc @bdhirsh

eellison replied:
There is also a `disable_meta` flag you could have applied that wouldn't introduce failures here

lezcano replied:
Sure, will do in the future.
It was not clear to me what the assert was, whether it was important or how to fix it. Adding more descriptive error messages would help.
